File: src/it/scala/com/exasol/cloudetl/kafka/docker/DockerNamedNetwork.scala
package com.exasol.cloudetl.kafka

import java.util.concurrent.ConcurrentHashMap

import com.typesafe.scalalogging.LazyLogging
import org.junit.runner.Description
import org.junit.runners.model.Statement
import org.testcontainers.DockerClientFactory
import org.testcontainers.containers.Network
import org.testcontainers.utility.ResourceReaper

/**
 * A reusable docker network.
 *
 * At the moment, the docker container {@code reuse} is ignored when a
 * network is attached to a container. This class creates docker network
 * that can be attached to reusable container.
 *
 * @param name name of the docker network
 * @param reuse boolean value to indicate reusability
 */
class DockerNamedNetwork(name: String, reuse: Boolean) extends Network with LazyLogging {

  private[this] val id = getNetworkId()

  override def getId(): String = id

  override def close(): Unit =
    if (reuse) {
      logger.warn(
        "Skipping the network termination because 'reuse' is enabled. Please destroy "
          + s"the network manually using 'docker network rm $id'."
      )
    } else {
      ResourceReaper.instance().removeNetworkById(id)
    }

  override def apply(base: Statement, description: Description): Statement =
    throw new UnsupportedOperationException()

  private[this] def getNetworkId(): String = {
    val network = DockerClientFactory
      .lazyClient()
      .listNetworksCmd()
      .withNameFilter(name)
      .exec()
      .stream()
      .findAny()

    if (network.isPresent()) {
      network.get().getId()
    } else {
      createNetwork()
    }
  }

  private[this] def createNetwork(): String =
    DockerClientFactory
      .lazyClient()
      .createNetworkCmd()
      .withName(name)
      .exec()
      .getId()
}

object DockerNamedNetwork {

  private[this] val namedNetworks: ConcurrentHashMap[String, DockerNamedNetwork] =
    new ConcurrentHashMap()

  def apply(name: String): DockerNamedNetwork =
    apply(name, false)

  def apply(name: String, reuse: Boolean): DockerNamedNetwork = {
    if (!namedNetworks.containsKey(name)) {
      namedNetworks.put(name, new DockerNamedNetwork(name, reuse))
    }
    namedNetworks.get(name)
  }
}

File: src/test/java/com/exasol/cloudetl/kafka/DockerNamedNetwork.java
package com.exasol.cloudetl.kafka;

import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Logger;

import org.testcontainers.DockerClientFactory;
import org.testcontainers.containers.Network;

/**
 * A reusable docker network.
 */
class DockerNamedNetwork implements Network {
    private static final Logger LOGGER = Logger.getLogger(DockerNamedNetwork.class.getName());
    private static final ConcurrentHashMap<String, DockerNamedNetwork> NAMED_NETWORKS = new ConcurrentHashMap<>();

    private final String name;
    private final boolean reuse;
    private final String id;

    static DockerNamedNetwork get(final String name, final boolean reuse) {
        return NAMED_NETWORKS.computeIfAbsent(name, key -> new DockerNamedNetwork(key, reuse));
    }

    private DockerNamedNetwork(final String name, final boolean reuse) {
        this.name = name;
        this.reuse = reuse;
        this.id = getNetworkId();
    }

    @Override
    public String getId() {
        return this.id;
    }

    @Override
    public void close() {
        if (this.reuse) {
            LOGGER.warning(() -> "Skipping the network termination because 'reuse' is enabled. Please destroy "
                    + "the network manually using 'docker network rm " + this.id + "'.");
        } else {
            DockerClientFactory.lazyClient().removeNetworkCmd(this.id).exec();
        }
    }

    private String getNetworkId() {
        return DockerClientFactory.lazyClient().listNetworksCmd().withNameFilter(this.name).exec().stream().findAny()
                .map(com.github.dockerjava.api.model.Network::getId).orElseGet(this::createNetwork);
    }

    private String createNetwork() {
        return DockerClientFactory.lazyClient().createNetworkCmd().withName(this.name).exec().getId();
    }
}
